src/mlpack/tests/main_tests/nmf_test.cpp

```diff
@@ -0,0 +1,294 @@
+ */
+BOOST_AUTO_TEST_CASE(NMFMultdivShapeTest)
+{
+  mat v = randu<mat>(10, 10);
+  int r = 5;
+
+  SetInputParam("update_rules", std::string("multdiv"));
+  SetInputParam("input", std::move(v));
+  SetInputParam("rank", r);
+
+  // Perform NMF.
+  mlpackMain();
+
+  // Get resulting matrices.
+  const mat& w = CLI::GetParam<mat>("w");
+  const mat& h = CLI::GetParam<mat>("h");
+
+  // Check the shapes of W and H.
+  BOOST_REQUIRE_EQUAL(w.n_rows, 10);
+  BOOST_REQUIRE_EQUAL(w.n_cols, 5);
+  BOOST_REQUIRE_EQUAL(h.n_rows, 5);
+  BOOST_REQUIRE_EQUAL(h.n_cols, 10);
+}
+
+/**
+ * Ensure the resulting matrices W, H have expected shape.
+ * Als update rule.
+ */
+BOOST_AUTO_TEST_CASE(NMFAlsShapeTest)
+{
+  mat v = randu<mat>(10, 10);
+  int r = 5;
+
+  SetInputParam("update_rules", std::string("als"));
+  SetInputParam("input", std::move(v));
+  SetInputParam("rank", r);
+
+  // Perform NMF.
+  mlpackMain();
+
+  // Get resulting matrices.
+  const mat& w = CLI::GetParam<mat>("w");
+  const mat& h = CLI::GetParam<mat>("h");
+
+  // Check the shapes of W and H.
+  BOOST_REQUIRE_EQUAL(w.n_rows, 10);
+  BOOST_REQUIRE_EQUAL(w.n_cols, 5);
+  BOOST_REQUIRE_EQUAL(h.n_rows, 5);
+  BOOST_REQUIRE_EQUAL(h.n_cols, 10);
+}
+
+/**
+ * Ensure the rank is positive.
+ */
+BOOST_AUTO_TEST_CASE(NMFRankBoundTest)
+{
+  mat v = randu<mat>(10, 10);
+  int r;
+
+  // Rank should not be negative.
+  r = -1;
+  SetInputParam("input", std::move(v));
+  SetInputParam("rank", r);
+ 
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+
+
+  // Rank should not be 0.
+  r = 0;
+  SetInputParam("rank", r);
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+/**
+ * Ensure the max_iterations is non-negative.
+ */
+BOOST_AUTO_TEST_CASE(NMFMaxIterartionBoundTest)
+{
+  mat v = randu<mat>(10, 10);
+  int r = 5;
+
+  // max_iterations should be non-negative.
+  SetInputParam("max_iterations", int(-1));
+  SetInputParam("input", std::move(v));
+  SetInputParam("rank", r);
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+/**
+ * Ensure the update rule is one of 
+ * {"multdist", "multdiv", "als"}.
+ */
+BOOST_AUTO_TEST_CASE(NMFUpdateRuleTest)
+{
+  mat v = randu<mat>(10, 10);
+  int r = 5;
+
+  // Invalid update rule.
+  SetInputParam("update_rules", std::string("invalid_rule"));
+  SetInputParam("input", std::move(v));
+  SetInputParam("rank", r);
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+/**
+ * Ensure min_residue is used, by testing that 
+ * min_resude makes a difference to the program.  
+ */
+BOOST_AUTO_TEST_CASE(NMFMinResidueTest)
+{
+  mat v = arma::randu(10, 10);
+  mat initial_w = arma::randu(10, 5);
+  mat initial_h = arma::randu(5, 10);
```

Comment:
Sorry for the nitpicking, but could you name these `initialW` and `initialH` in accordance with the style guide?  (variable names shouldn't have underscores)